### ISSUE TITLE
Error in `integration with uniswap3 core` test: `Timeout of 20000ms exceeded.`

### ISSUE THREAD

**Issue by cameel** [euler-xyz/euler-contracts#97]
Here's another flaky test:

```
 298 passing (10m)
  1 failing

  1) uniswap3 twap
       integration with uniswap3 core:
     Error: Timeout of 20000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/tmp/ext-test-Euler-X7BOE6/ext/test/uniswap-integration.js)
      at listOnTimeout (node:internal/timers:557:17)
      at processTimers (node:internal/timers:500:7)
```
I've seen several instances of it in the last 3 days: [966585](https://app.circleci.com/pipelines/github/ethereum/solidity/22038/workflows/caa95319-bae0-4f82-9bc4-f1161cbdfc35/jobs/966585), [966481](https://app.circleci.com/pipelines/github/ethereum/solidity/22037/workflows/f460c998-2164-4d97-848c-4624732e3ac5/jobs/966481), [965865](https://app.circleci.com/pipelines/github/ethereum/solidity/22023/workflows/091aefd3-2773-4fc3-9591-0db04ca1a49a/jobs/965865), [965823](https://app.circleci.com/pipelines/github/ethereum/solidity/22022/workflows/14f1683f-1d8c-4062-9f60-7e6bce1d061c/jobs/965823), [965567](https://app.circleci.com/pipelines/github/ethereum/solidity/22016/workflows/58a14e67-4df7-41ec-bb90-36309c63c1a2/jobs/965567).

Does not seem to be a race condition like in #1. Or at least `TEST_SLEEP` does not make it easily reproducible - I've only seen it in CI so far and not in local runs.

**Comment by cameel** [comment]
I have just seen this happen in another test ([968714](https://app.circleci.com/pipelines/github/ethereum/solidity/22082/workflows/200584c7-1c4f-43ad-b429-cb0e2e852034/jobs/968714)):
```
  309 passing (8m)
  1 failing

  1) EUL distributor
       merkle distribution of EUL token:
     Error: Timeout of 20000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/tmp/ext-test-Euler-bpGAjC/ext/test/eulDistributor.js)
      at listOnTimeout (node:internal/timers:557:17)
      at processTimers (node:internal/timers:500:7)
```
Before it was always in `uniswap3 twap`.

**Comment by hoytech** [comment]
I think these are just legitimately long-running tests, so we have increased their timeouts. At some point we should try to improve our test performance!

**Comment by cameel** [comment]
Ah, ok. Sounds fine to me. 

Explains why I'm seeing it in other projects sometimes - I'm actually investigating [another one of this kind right now, this time in PRBMath](https://app.circleci.com/pipelines/github/ethereum/solidity/22165/workflows/590bfab7-3e2e-47c4-aaba-49bb55ccf483/jobs/972388).

**Comment by cameel** [comment]
Here's another one that might need adjustment ([975728](https://app.circleci.com/pipelines/github/ethereum/solidity/22237/workflows/acda36c8-f05b-44b4-ba20-278f97b7c7a4/jobs/975728)):
```
 1) burn
       burn with max_uint256 repays the debt in full or up to the available underlying balance:
     Error: Timeout of 20000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/tmp/ext-test-Euler-2z6LHr/ext/test/burn.js)
      at listOnTimeout (node:internal/timers:557:17)
      at processTimers (node:internal/timers:500:7)

```

By the way, @paulrberg mentioned in https://github.com/paulrberg/prb-math/issues/71 that Hardhat changed the default timeout from 30 sec to 20. If your tests were created with that longer timeout in mind maybe it would be a good idea to set it to this older value globally?

**Comment by cameel** [comment]
More timeouts require adjustments.

[978941](https://app.circleci.com/pipelines/github/ethereum/solidity/22304/workflows/05fb5262-0b44-4822-ba3d-7c1510f2d394/jobs/978941):
```
309 passing (17m)
  1 failing

  1) views
       basic view, APR/APY:
     Error: Timeout of 20000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/tmp/ext-test-Euler-KI4V3v/ext/test/view.js)
      at listOnTimeout (node:internal/timers:557:17)
      at processTimers (node:internal/timers:500:7)
```

[978926](https://app.circleci.com/pipelines/github/ethereum/solidity/22303/workflows/0ec09443-caa0-4bac-b719-7551c1acb318/jobs/978926):
```
  309 passing (18m)
  1 failing

  1) activating markets
       re-activate:
     Error: Timeout of 20000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/tmp/ext-test-Euler-OOjyIx/ext/test/activateMarket.js)
      at listOnTimeout (node:internal/timers:557:17)
      at processTimers (node:internal/timers:500:7)
```

By the way, are you seeing these in your testing too? I wonder if I should just set a bigger timeout in settings in our tests.

**Comment by kasperpawlowski** [comment]
@cameel could you please try running the tests using `timeout-fix` branch? I increased the global timeout for the tests to 100 sec and hopefully it will fix those issues. We do not see them at our end.

https://github.com/euler-xyz/euler-contracts/tree/timeout-fix

**Comment by cameel** [comment]
Thanks! I'll let you know if that helps. This is happening regularly and there are even mass failures like [1000528](https://app.circleci.com/pipelines/github/ethereum/solidity/22791/workflows/7b3fc28b-881b-4b50-aa4f-b47c55d4261b/jobs/1000528) where 3 or more tests fail in the same run due to timeouts.

But I think I found potential culprit. I went through some of the failed runs and the common patterns is that it's always the run without optimization. So it's possible that your timeouts are just enough for optimized code but are too short in a general case.

If that's the case and you don't care about builds without optimization, you might want to keep the timeout as is in your repo and I could just override it on our side.

**Comment by kasperpawlowski** [comment]
@cameel you're right. to be honest I'm new to euler and didn't expect that the global mocha timeout setting might be overridden.  the following PR should fix that

https://github.com/euler-xyz/euler-contracts/pull/112
